### PR TITLE
Forcing done right

### LINKED ISSUE #3544: internal error @ TypeChecking/Forcing.hs:227
Using the latest version (master, 449d40be5d89d77bc7a119865ab5c8bda56b5fe9), Agda reports an internal error:

```agda
data Nat : Set where
  succ : Nat → Nat

data Fin : Nat → Set where
  zero : (n : Nat) → Fin (succ n)

data Tm (n : Nat) : Set where
  var : Fin n → Tm n
  piv : Fin (succ n) → Tm n

data Cx : Nat → Set where
  succ : (n : Nat) → Tm n → Cx (succ n)

data CxChk : ∀ n → Cx n → Set where
  succ : (n : Nat) (T : Tm n) → CxChk (succ n) (succ n T)

data TmChk (n : Nat) : Cx n → Tm n → Set where
  vtyp : (g : Cx n) (v : Fin n) → CxChk n g → TmChk n g (var v)

error : ∀ n g s → TmChk n g s → Set
error n g s (vtyp g' (zero x) (succ n' (piv (zero y)))) = Nat -- Internal error here.
error _ _ _ (vtyp .(succ n (var x)) (zero n) (succ .n (var x))) = Nat -- This clause added to pass 2.5.3.
```

The report is:
```
An internal error has occurred. Please report this as a bug.
Location of the error: src/full/Agda/TypeChecking/Forcing.hs:227
```

### LINKED ISSUE #3903: Forcing translation produces segfaulting program
While trying to understand the forcing translation with @jespercockx, we found out that it actually does not work.  The forcing translation tries to relocate a match happening in a forced position to a non-forced position.  In the example below, we have a match `cons zero ...` but pattern `zero` is in a forced position.  The forcing translation now tries to reconstruct a match against `zero` by transforming (parts of) a dot pattern into an actual pattern.  However, it blindly grabs the first opportunity.  In the example below, there are two possibilities, and Agda takes the wrong one, leading to a segfaulting compiled code.  And in fact, without type information there is no way for Agda to decide which is the correct one.  So, if it works, it is pure luck.
```agda
open import Agda.Builtin.Unit
open import Agda.Builtin.Bool
open import Agda.Builtin.Nat
open import Agda.Builtin.Equality
open import Agda.Builtin.IO
open import Common.String

postulate
  putStr : String → IO ⊤

{-# FOREIGN GHC import qualified Data.Text.IO #-}
{-# COMPILE GHC putStr = Data.Text.IO.putStr #-}

data Sing : Nat → Set where
  sing : Sing zero

data Vec : Nat → Set where
  nil : Vec zero
  cons : ∀ n → Vec n → Vec (suc n)

isTailNil : ∀ m n → Sing m → Vec n → Bool
isTailNil .0 .1 sing (cons zero v) = true  -- two `zero`s are buried in the dot patterns, which to resurrect?
isTailNil _ _ _ _ = false

shouldBeFalse = isTailNil 0 2 sing (cons 1 (cons 0 nil))

isFalse : false ≡ shouldBeFalse
isFalse = refl

magic : false ≡ true → String
magic ()

f : (b : Bool) → false ≡ b → String
f false eq = "Phew!"
f true  eq = magic eq

main = putStr (f shouldBeFalse refl)
```
The compiled program crashes with:
```
FoolForcing: Agda: unreachable code reached.
CallStack (from HasCallStack):
  error, called at ./MAlonzo/RTE.hs:46:23 in main:MAlonzo.RTE
```
At least for this example, this seems to be a regression.  It worked correctly in Agda-2.5.4.2.  (How did previous Agda do this forcing translation?)

From Agda-2.6.0, the forcing translation produces a match on the wrong (first) argument of `d20`.
```haskell
name20 = "FoolForcing.isTailNil"
d20 :: Integer -> Integer -> T4 -> T8 -> Bool
d20 v0 v1 v2 v3 = du20 v0 v3
du20 :: Integer -> T8 -> Bool
du20 v0 v1
  = let v2 = MAlonzo.Code.Agda.Builtin.Bool.C8 in
    case coe v1 of
      C14 v4
        -> case coe v0 of
             0 -> coe MAlonzo.Code.Agda.Builtin.Bool.C10
             _ -> coe v2
      _ -> coe v2
```
Agda-2.5.4.2 produces the correct code, matching on the second argument of `d20`:
```haskell
name20 = "FoolForcing.isTailNil"
d20 :: Integer -> Integer -> T4 -> T8 -> Bool
d20 v0 v1 v2 v3 = du20 v1 v3
du20 :: Integer -> T8 -> Bool
du20 v0 v1
  = let v2 = MAlonzo.Code.Agda.Builtin.Bool.C8 in
    case coe v1 of
      C14 v4
        -> case coe v0 of
             1 -> coe MAlonzo.Code.Agda.Builtin.Bool.C10
             _ | coe (geqInt (coe v0) (coe (1 :: Integer))) -> coe v2
             _ -> coe v2
      _ -> coe v2
```

### LINKED ISSUE #3930: BUILTIN NATURAL internal error at Forcing.hs:232
Checking this example in Agda 2.6.0.1
```agda
module issue3930 where

  data ℕ : Set where
    zero : ℕ
    suc : ℕ → ℕ
  {-# BUILTIN NATURAL ℕ #-}

  _+_ : (m n : ℕ) → ℕ
  zero + n = n
  suc m + n = suc (m + n)

  data Th : (m n : ℕ) → Set where
    os : ∀ {m n} → Th m n → Th (suc m) (suc n)

  Fin : ℕ → Set
  Fin = Th (suc zero)

  infixl 6 _+↓_
  infix 4 _≈M_

  postulate
    U : Set
    RCtx : ℕ → Set
    _+↓_ : ∀ {m n} → RCtx m → RCtx n → RCtx (n + m)
    [-_-] : U → RCtx (suc zero)
    _≈M_ : ∀ {n} (Δ0 Δ1 : RCtx n) → Set

  infix 4 _⊢l-var_

  data _⊢l-var_ : ∀ {n} (Δi : RCtx n) (i : Fin n) → Set where
    os : ∀ {n} {e : Th zero n} {Δ Δπ π} (iq : Δπ ≈M Δ +↓ [- π -]) →
         Δπ ⊢l-var os e

  ⊢l-var-sub : ∀ {n Δ} {i : Fin n} → Δ ⊢l-var i → Set
  ⊢l-var-sub (os iq) = {!!}
```
gives error message
```
An internal error has occurred. Please report this as a bug.
Location of the error: src/full/Agda/TypeChecking/Forcing.hs:232
```

Note that the error only turns up at the pattern match, not at the data declaration.

Some experiments:

* Removing the pragma `  {-# BUILTIN NATURAL ℕ #-}` makes it check fine.
* Replacing all of the expressions like `Δ +↓ [- π -]` using a new postulate `_-,_ : RCtx m → U → RCtx (suc m)`, resulting in `Δ -, π` makes it check fine.
* Using `{-# BUILTIN NATPLUS _+_ #-}` makes no difference.
* Removing the index `i : Fin n` makes everything check fine.

### LINKED ISSUE #4081: Forcing translation crashes
```agda
open import Agda.Builtin.Equality

module _ (A : Set) where

data Wrap (X : Set) : Set where
  wrap : X → Wrap X

data D (x : A) : ∀ y → Wrap (x ≡ y) → Set where
  c : ∀ y (x≡y : x ≡ y) → D x y (wrap x≡y)

test : ∀ x y (x≡y x≡y' : Wrap (x ≡ y)) → D x y x≡y → D x y x≡y' → Set
test y .y (wrap refl) .(wrap refl) (c .y .refl) (c .y refl) = A
```
produces
```
An internal error has occurred. Please report this as a bug.
Location of the error: src/full/Agda/TypeChecking/Forcing.hs:241
```

### LINKED ISSUE #4116: Internal error Forcing.hs:232
When I press `C-c C-c x` on the hole in the example below, the argument is correclty expanded into cases, but instead of creating new holes for each case, I get the following error:
```
An internal error has occurred. Please report this as a bug.
Location of the error: src/full/Agda/TypeChecking/Forcing.hs:232
```

```
Agda version 2.6.0.1
```

Example code:
```agda
open import Data.Nat
open import Data.Fin hiding (_+_)
open import Data.Vec

data Ar {a} (X : Set a) (d : ℕ) (s : Fin d → ℕ) : Set a where
  imap : (((i : Fin d) → Fin (s i)) → X) → Ar X d s

module _ where 
  data Tensor {a} : (X : Set a) → (n : ℕ) → (s : Vec ℕ n) → Set a where
    zz : ∀ {X : Set a} → (x : X) → Tensor X 0 []
    ss : ∀ {X : Set a}{n m}{s : Vec ℕ n} → (v : Vec (Tensor X n s) m) → Tensor X (1 + n) (m ∷ s)

  tensor→a : ∀ {a}{X : Set a}{n s} → Tensor X n s → Ar X n (lookup s)
  tensor→a x = {!!}
```
I have seen two open bugs mentioning the same internal error, but it is difficult to tell whether it points to the same problem.

### PR DESCRIPTION
**Pull request by UlfNorell**
Moves the forcing translation to the lhs unifier. This is arguably the correct place for it and has the benefit that we never work with clauses (compiled or otherwise) that have not been forcing-translated.

The basic idea is that when the unifier wants to solve a constraint `x = c vs` where `vs` contains some forced variable it produces a lazy pattern binding the forced variable, instead of (as before) producing a dot pattern.

Fixes #3544
Fixes #3903
Fixes #3930
Fixes #4081
Fixes #4116

### CODE REVIEW

**Review by jespercockx — APPROVED** [review]

**Comment by UlfNorell** [comment]
Test failure is due to #4179 and not introduced by this PR.

File: src/full/Agda/TypeChecking/Monad/Base.hs [hunk 1]
```diff
@@ -2555,6 +2558,7 @@ data TCEnv =
           , envMutualBlock         :: Maybe MutualId -- ^ the current (if any) mutual block
           , envTerminationCheck    :: TerminationCheck ()  -- ^ are we inside the scope of a termination pragma
           , envCoverageCheck       :: CoverageCheck        -- ^ are we inside the scope of a coverage pragma
+          , envMakeCase            :: Bool                 -- ^ are we inside a make-case (if so, ignore forcing analysis in unifier)
```
Comment (andreasabel) [inline src/full/Agda/TypeChecking/Monad/Base.hs:2561]: The style guide says:  Don't use Bool!
How about:
```haskell
data InsideMakeCase
  = InsideMakeCase
  | NotInsideMakeCase
```

File: src/full/Agda/TypeChecking/Patterns/Match.hs [hunk 2]
```diff
@@ -55,24 +57,25 @@ buildSubstitution :: (DeBruijn a)
...
+    -- meta-variables will make progress.
+    DontKnow b <> DontKnow b'      = DontKnow $ b <> b'
 
---     -- @NotBlocked (StuckOn e)@ means blocked by a variable.
---     -- In this case, no instantiation of
---     -- meta-variables will make progress.
---     DontKnow b `mappend` DontKnow b'      = DontKnow $ b `mappend` b'
+    -- One could imagine DontKnow _ <> No = No, but would break the
+    -- equivalence to case-trees.
+    DontKnow m <> _                = DontKnow m
```
Comment (andreasabel) [inline src/full/Agda/TypeChecking/Patterns/Match.hs:64]: This pattern match could be reordered for clarity: `DontKnow` is dominant, so put the cases involving `DontKnow` on top (3 cases).  Next in the pecking order is `No` (2 cases).  Last is `Yes` (1 case).

**Review by andreasabel — COMMENTED** [review]
Does this also fix #2640, case `test/Compiler/simple/Erasue-Issue2640`?  This test case is disabled still...

**Comment by UlfNorell** [comment]
> Does this also fix #2640, case test/Compiler/simple/Erasue-Issue2640? This test case is disabled still...

It should, but for some reason it doesn't. I'll investigate.

**Comment by UlfNorell** [comment]
It looks like we don't properly keep track of modalities in @-bindings generated by the unifier. I'll file a new issue once this is merged.

**Comment by UlfNorell** [comment]
Issue for modalities here: #4195

**Pull request merged**